File: telegram-service/src/main/java/de/nak/home_assistant/actions/DeviceList.java
package de.nak.home_assistant.actions;

import com.google.common.collect.Lists;
import com.pengrad.telegrambot.request.SendMessage;
import de.nak.home_assistant.CustomKeyboard;
import de.nak.home_assistant.models.service.devices.Device;
import de.nak.home_assistant.models.service.DeviceListResponse;
import de.nak.home_assistant.models.database.ModuleResponse;
import de.nak.home_assistant.services.ServiceService;
import de.nak.home_assistant.services.database.UserService;

import java.util.List;

public class DeviceList {

    public static SendMessage generateMessage(long userId, long chatId, int moduleId) {
        // Make initial User requests
        UserService userService = new UserService(userId);
        ModuleResponse moduleResponse = userService.getUserModule(moduleId);

        // Request DeviceList from Service
        ServiceService service = new ServiceService(new DeviceListResponse());
        DeviceListResponse deviceListResponse = service.makeRequest(moduleResponse.getServiceRequest());

        SendMessage message;

        if (deviceListResponse.isSuccess()) {

            // Partition the DeviceList so each row has 3 Buttons max
            List<List<Device>> devices = Lists.partition(deviceListResponse.getResult(), 3);

            message = new SendMessage(chatId, "Deine " + moduleResponse.getName() + " Geräte:")
                    .replyMarkup(CustomKeyboard.generateDeviceButtons(devices, moduleId));
        } else {
            message = new SendMessage(chatId, "Etwas hat nicht funktioniert: " + deviceListResponse.getError());
        }

        return message;
    }

}

File: telegram-service/src/actions/device-list.ts
import bot from "../bot";
import { i18n } from "../i18n";
import { generateDeviceButtons } from "../keyboard";
import { makeServiceRequest } from "../services/service";
import { DeviceListResponse } from "../services/typings";
import { getUserModule } from "../services/user";

export const sendDeviceList = async (
  userId: number,
  chatId: number,
  moduleId: number
) => {
  const module = await getUserModule(userId, moduleId);
  const translations = await i18n(userId);

  if (module) {
    const {
      success,
      result,
      error,
    } = await makeServiceRequest<DeviceListResponse>(module.serviceRequest);

    if (success) {
      const partitionSize = 3;

      // Partition the DeviceList so each row has 3 Buttons max
      const devices = new Array(Math.ceil(result.length / partitionSize))
        .fill([])
        .map(() => result.splice(0, partitionSize));

      return bot.sendMessage(
        chatId,
        translations.deviceList.title(module.name),
        {
          reply_markup: generateDeviceButtons(devices, moduleId),
        }
      );
    }

    return bot.sendMessage(chatId, `${translations.internalError}: ${error}`);
  }

  return bot.sendMessage(chatId, translations.deviceList.error);
};
